src/indexer.c

```diff
@@ -126,59 +130,113 @@ int CountPredicateType(QueryPredicateList *queries, PredicateType type) {
+        RedisModuleString *indexed_key = RedisModule_CreateStringPrintf(NULL, K_PREFIX, key_string);
 
-        indexUnderKey(op, indexed_key_value, ts_key);
-        indexUnderKey(op, indexed_key, ts_key);
+        labelIndexUnderKey(Indexer_Add, indexed_key_value, ts_key, labelsIndex, tsLabelIndex);
+        labelIndexUnderKey(Indexer_Add, indexed_key, ts_key, labelsIndex, tsLabelIndex);
 
-        RedisModule_FreeString(ctx, indexed_key_value);
-        RedisModule_FreeString(ctx, indexed_key);
+        RedisModule_FreeString(NULL, indexed_key_value);
+        RedisModule_FreeString(NULL, indexed_key);
     }
 }
 
-void IndexMetric(RedisModuleCtx *ctx,
-                 RedisModuleString *ts_key,
-                 Label *labels,
-                 size_t labels_count) {
-    IndexOperation(ctx, Indexer_Add, ts_key, labels, labels_count);
+// Removes the ts from the label index and from the inverse index, if exist.
+// del_key should be false if caller wants to avoid iterator invalidation.
+void RemoveIndexedMetric_generic(RedisModuleString *ts_key,
+                                 RedisModuleDict *_labelsIndex,
+                                 RedisModuleDict *_tsLabelIndex,
+                                 bool del_key) {
+    int nokey = 0;
+    RedisModuleDict *ts_leaf = RedisModule_DictGet(_tsLabelIndex, ts_key, &nokey);
+    if (nokey) { // series has no labels or already been removed from index
+        return;
+    }
+
+    RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(ts_leaf, "^", NULL, 0);
+    RedisModuleString *currentLabelKey;
+    while ((currentLabelKey = RedisModule_DictNext(NULL, iter, NULL)) != NULL) {
+        labelIndexUnderKey(Indexer_Remove, currentLabelKey, ts_key, _labelsIndex, _tsLabelIndex);
+        RedisModule_FreeString(NULL, currentLabelKey);
+    }
+    RedisModule_DictIteratorStop(iter);
+    RedisModule_FreeDict(NULL, ts_leaf);
+    if (del_key) {
+        RedisModule_DictDel(_tsLabelIndex, ts_key, NULL);
+    }
+}
+
+// Removes the ts from the label index and from the inverse index, if exist.
+void RemoveIndexedMetric(RedisModuleString *ts_key) {
+    RemoveIndexedMetric_generic(ts_key, labelsIndex, tsLabelIndex, true);
 }
 
-void RemoveIndexedMetric(RedisModuleCtx *ctx,
-                         RedisModuleString *ts_key,
-                         Label *labels,
-                         size_t labels_count) {
-    IndexOperation(ctx, Indexer_Remove, ts_key, labels, labels_count);
+// Removes all indexed metrics
+void RemoveAllIndexedMetrics_generic(RedisModuleDict *_labelsIndex,
+                                     RedisModuleDict **_tsLabelIndex) {
+    RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(*_tsLabelIndex, "^", NULL, 0);
+    RedisModuleString *currentTSKey;
+    while ((currentTSKey = RedisModule_DictNext(NULL, iter, NULL)) != NULL) {
+        RemoveIndexedMetric_generic(currentTSKey, _labelsIndex, *_tsLabelIndex, false);
```

Comment:
same as RemoveIndexedMetric_generic

Reply:
This function iterates over all the keys and removes them from the index by calling RemoveIndexedMetric_generic.